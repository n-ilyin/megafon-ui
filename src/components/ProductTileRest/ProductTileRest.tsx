import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './ProductTileRest.less';
import Header from '../Header/Header';
import TextLink from '../TextLink/TextLink';
import Button from '../Button/Button';
import DropdownSocialList from '../DropdownSocialList/DropdownSocialList';

interface IProductTileRestProps {
    /** Tile title */
    title: string;
    /** Tile description */
    description?: string;
    /** Shop tag */
    shopTag?: string;
    /** Link more - href */
    link?: string;
    /** Link more - text */
    moreLinkText?: string;
    /** Link more - show/hide */
    showMoreLink?: boolean;
    /** Button buy - href */
    buyLink?: string;
    /** Button buy - text */
    buyButtonText?: string;
    /** Button buy - show/hide */
    showBuyButton?: boolean;
    /** Button buy - passive color */
    buttonPassiveColor?: 'green' | 'transparent-green';
    /** Link connect - href */
    connectLink?: string;
    /** Link connect - text */
    connectButtonText?: string;
    /** Link connect - show/hide */
    showConnectButton?: boolean;
    /** Payment */
    payment: {
        title: string;
        value: string;
        unitExtra: string;
        unitValue: string;
        discount?: string;
    };
    /** Packs */
    packs: Array<{
        value: number;
        unit: string;
        title: string;
        isDelim: boolean;
    }>;
    /** First params */
    firstParams: {
        title?: string;
        caption: string;
        items: Array<{
            title: string;
            svgIcon: JSX.Element;
        }>;
    };
    /** Second params */
    secondParams: Array<{
        title: string;
        value?: string;
        unit?: string;
    }>;
    /** Info - object with custom properties
     * - return with onClickConnect, onClickBuy, onClickMore, onClickBubble
     */
    info: {};
    /** Button connect handler - return object { ...info, callsValue, trafficValue } */
    onClickConnect?(data: {}, e: React.SyntheticEvent<EventTarget>): void;
    /** Button buy handler - return object
     * { ...info, shopTag }
     */
    onClickBuy?(data: {}, e: React.SyntheticEvent<EventTarget>): void;
    /** Link more handler - return object
     * { ...info, shopTag }
     */
    onClickMore?(data: {}, e: React.SyntheticEvent<EventTarget>): void;
}

const cn = cnCreate('mfui-product-tile-rest');
class ProductTileRest extends React.Component<IProductTileRestProps> {
    static propTypes = {
        title: PropTypes.string.isRequired,
        description: PropTypes.string,
        shopTag: PropTypes.string,
        link: PropTypes.string,
        moreLinkText: PropTypes.string,
        showMoreLink: PropTypes.bool,
        buyLink: PropTypes.string,
        buyButtonText: PropTypes.string,
        showBuyButton: PropTypes.bool,
        connectLink: PropTypes.string,
        connectButtonText: PropTypes.string,
        showConnectButton: PropTypes.bool,
        buttonPassiveColor: Button.propTypes.passiveColor,
        payment: PropTypes.shape({
            title: PropTypes.string,
            value: PropTypes.string,
            unitExtra: PropTypes.string,
            unitValue: PropTypes.string,
            discount: PropTypes.string,
        }).isRequired,
        packs: PropTypes.arrayOf(PropTypes.shape({
            value: PropTypes.number,
            unit: PropTypes.string,
            title: PropTypes.string,
            isDelim: PropTypes.bool,
        })).isRequired,
        firstParams: PropTypes.shape({
            title: PropTypes.string,
            caption: PropTypes.string,
            items: PropTypes.arrayOf(PropTypes.shape({
                title: PropTypes.string,
                svgIcon: PropTypes.element,
            })),
        }).isRequired,
        secondParams: PropTypes.arrayOf(PropTypes.shape({
            title: PropTypes.string,
            value: PropTypes.string,
            unit: PropTypes.string,
        })).isRequired,
        info: PropTypes.object,
        onClickConnect: PropTypes.func,
        onClickBuy: PropTypes.func,
        onClickMore: PropTypes.func,
    };

    static defaultProps: Partial<IProductTileRestProps> = {
        moreLinkText: 'Подробнее',
        showMoreLink: true,
        buyButtonText: 'Купить',
        showBuyButton: true,
        buttonPassiveColor: 'green',
        connectButtonText: 'Перейти на тариф',
        showConnectButton: true,
        shopTag: '',
    };

    handleClickConnect = (e: React.SyntheticEvent<EventTarget>) => {
        const { info, shopTag, onClickConnect } = this.props;

        onClickConnect && onClickConnect({ ...info, shopTag }, e);
    }

    handleClickBuy = (e: React.SyntheticEvent<EventTarget>) => {
        const {
            info,
            shopTag,
            onClickBuy,
            payment: {
                value,
                discount,
                unitValue,
                unitExtra,
            },
        } = this.props;
        const priceValue = discount || value;

        onClickBuy && onClickBuy({
            ...info,
            shopTag,
            price: priceValue,
            unitValue,
            unitExtra,
        }, e);
    }

    handleClickMore = (e: React.SyntheticEvent<EventTarget>) => {
        const { info, shopTag, onClickMore } = this.props;

        onClickMore && onClickMore({ ...info, shopTag }, e);
    }

    renderIcons() {
        const { firstParams: { title, caption, items } } = this.props;

        return (
            <div className={cn('messangers')}>
                {!!items.length &&
                    <div className={cn('messangers-list')}>
                        <DropdownSocialList
                            icons={items}
                            maxIconNumber={4}
                            className={cn('options-list')}
                        />
                    </div>
                }
                {title &&
                    <div className={cn('messangers-description')}>
                        {title}<br /> {caption}
                    </div>
                }
            </div>
        );
    }

    renderOptions() {
        const { secondParams } = this.props;

        if (!secondParams.length) {
            return null;
        }

        return (
            <div className={cn('options')}>
                {secondParams.map((param, index) => {
                    const { title, value, unit } = param;

                    return (
                        <div className={cn('option')} key={title + index}>
                            <div className={cn('option-title')}>{title}</div>
                            <div className={cn('option-description')}>{`${value} ${unit}`}</div>
                        </div>
                    );
                })}
            </div>
        );
    }

    renderShowcase() {
        const {
            payment: {
                title,
                value,
                unitExtra,
                unitValue,
                discount,
            },
            packs,
        } =  this.props;

        return (
            <React.Fragment>
                <div className={cn('price', { discount: !!discount })}>
                    <div className={cn('discount-condition')}>{title}</div>
                    <div className={cn('old-price-wrapper')}>
                        <div className={cn('old-price')}>{`${value} ${unitValue}`}</div>
                    </div>
                    <div className={cn('actual-price')}>{`${discount || value} ${unitValue} ${unitExtra}`}</div>
                </div>
                {packs.map((param, index) => {
                    const {
                        value: paramValue,
                        unit: paramUnit,
                        title: paramTitle,
                        isDelim,
                    } = param;

                    return (
                        <div
                            className={cn('pack')}
                            key={paramValue + paramUnit + index}
                        >
                            {isDelim ? paramTitle : `${paramValue} ${paramUnit}`}
                        </div>
                    );
                })}
            </React.Fragment>
        );
    }

    render() {
        const {
            title,
            description,
            link,
            showMoreLink,
            moreLinkText,
            connectLink,
            connectButtonText,
            showConnectButton,
            buyLink,
            showBuyButton,
            buyButtonText,
            buttonPassiveColor,
        } = this.props;

        return (
            <div className={cn('')}>
                <div className={cn('info')}>
                    <Header className={cn('header')} as="h3">{title}</Header>
                    {showMoreLink &&
                        <TextLink
                            className={cn('detail-link', { type: 'more' })}
                            href={link}
                            target="_blank"
                            onClick={this.handleClickMore}
                        >
                            {moreLinkText}
                        </TextLink>
                    }
                    {this.renderShowcase()}
                    {this.renderIcons()}
                    {this.renderOptions()}
                    {description &&
                        <div className={cn('description')}>{description}</div>
                    }
                </div>
                <div className={cn('buy')}>
                    {showBuyButton &&
                        <Button
                            className={cn('buy-button', { 'additional-margin': !showConnectButton })}
                            passiveColor={buttonPassiveColor}
                            hoverColor="green"
                            sizeAll="medium"
                            href={buyLink}
                            onClick={this.handleClickBuy}
                        >
                            {buyButtonText}
                        </Button>
                    }
                    {showConnectButton &&
                        <TextLink
                            className={cn('detail-link', { type: 'connect' })}
                            href={connectLink}
                            onClick={this.handleClickConnect}
                            target="_blank"
                        >
                            {connectButtonText}
                        </TextLink>
                    }
                </div>
            </div>
        );
    }
}

export default ProductTileRest;
