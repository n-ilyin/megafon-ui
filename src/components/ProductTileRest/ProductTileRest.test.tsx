import * as React from 'react';
import { shallow, mount } from 'enzyme';
import ProductTileRest from './ProductTileRest';
import MegafonTv from 'icons/Service-logos/24/MF-TV.svg';

const className = '.mfui-product-tile-rest';

const info = {
    billingIds: [ '1223' ],
    fast: false,
    sms: null,
    smsText: null,
    smsSite: null,
    ivr: null,
    ussd: null,
    lkLink: 'https://lk.megafon.ru/tariffs/',
    popupText: null,
    title: 'Включайся! Выбирай',
    id: 2098,
};

const tariff = {
    title: 'Включайся! Выбирай',
    description: 'description',
    shopTag: 'choose',
    link: '/tariffs/all/vybiray.html',
    moreLinkText: 'Подробнее',
    showMoreLink: true,
    isBuyable: true,
    buyLink: '/zakaz/?tariff=choose',
    buyButtonText: 'Купить',
    showBuyButton: true,
    buttonPassiveColor: 'green' as 'green',
    connectLink: '/tariffs/all/vybiray.html',
    connectButtonText: 'Подключить',
    showConnectButton: true,
    payment: {
        title: 'При покупке новой SIM–карты',
        value: '3000',
        unitExtra: 'за 30 дней',
        unitValue: '₽',
        discount: '100',
    },
    packs: [
        {
            value: 10,
            unit: 'ГБ',
            title: 'Звонки на все номера России',
            isDelim: false,
        },
        {
            value: 10,
            unit: 'ГБ',
            title: 'Безлимитный интернет',
            isDelim: true,
        },
    ],
    firstParams: {
        title: 'Параметры',
        caption: 'Тариф для планшетов',
        items: [{
            svgIcon: <MegafonTv />,
            title: 'МегаФон ТВ',
        }],
    },
    secondParams: [
        {
            title: 'Кэшбэк 20%',
            unit: '',
            value: '',
        },
        {
            title: 'Безлимитные SMS',
            unit: '₽ за 30 дней',
            value: '100',
        },
    ],
    info,
};

const handler = jest.fn();

beforeEach(() => {
    jest.clearAllMocks();
});

describe('<ProductTileRest />', () => {
    describe('render', () => {
        it('renders ProductTileRest with empty props', () => {
            const wrapper = shallow(
                <ProductTileRest
                    title="title"
                    payment={{
                        title: 'title',
                        value: 'value',
                        unitExtra: 'unitExtra',
                        unitValue: 'unitValue',
                    }}
                    packs={[]}
                    firstParams={{
                        caption: 'caption',
                        items: [],
                    }}
                    secondParams={[]}
                    info={{}}
                />
            );

            expect(wrapper).toMatchSnapshot();
        });

        it('renders ProductTileRest', () => {
            const wrapper = shallow(<ProductTileRest {...tariff} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    it('calls onClickConnect', () => {
        const wrapper = shallow(
            <ProductTileRest
                {...tariff}
                info={info}
                onClickConnect={handler}
            />
        );
        const connectButton = `${className}__detail-link_type_connect`;
        const onClick = wrapper.find(connectButton).prop('onClick') as any;

        onClick({});
        expect(handler).toBeCalledWith({
            ...info,
            shopTag: tariff.shopTag,
        }, {});
    });

    it('calls onClickBuy', () => {
        const wrapper = shallow(
            <ProductTileRest
                {...tariff}
                info={info}
                onClickBuy={handler}
            />
        );
        const buyButton = `${className}__buy-button`;
        const onClick = wrapper.find(buyButton).prop('onClick') as any;

        onClick({});
        expect(handler).toBeCalledWith({
            ...info,
            shopTag: tariff.shopTag,
            price: tariff.payment.discount,
            unitValue: tariff.payment.unitValue,
            unitExtra: tariff.payment.unitExtra,
        }, {});
    });

    it('calls onClickMore', () => {
        const wrapper = shallow(
            <ProductTileRest
                {...tariff}
                info={info}
                onClickMore={handler}
            />
        );
        const moreButton = `${className}__detail-link_type_more`;
        const onClick = wrapper.find(moreButton).prop('onClick') as any;

        onClick({});
        expect(handler).toBeCalledWith({
            ...info,
            shopTag: tariff.shopTag,
        }, {});
    });
});
