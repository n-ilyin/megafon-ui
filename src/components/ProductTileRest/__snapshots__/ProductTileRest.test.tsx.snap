// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTileRest /> render renders ProductTileRest 1`] = `
<div
  className="mfui-product-tile-rest"
>
  <div
    className="mfui-product-tile-rest__info"
  >
    <Header
      as="h3"
      className="mfui-product-tile-rest__header"
      color="black"
    >
      Включайся! Выбирай
    </Header>
    <TextLink
      className="mfui-product-tile-rest__detail-link mfui-product-tile-rest__detail-link_type_more"
      color="blue"
      href="/tariffs/all/vybiray.html"
      onClick={[Function]}
      target="_blank"
      underlineStyle="solid"
      underlineVisibility="hover"
    >
      Подробнее
    </TextLink>
    <div
      className="mfui-product-tile-rest__price mfui-product-tile-rest__price_discount"
    >
      <div
        className="mfui-product-tile-rest__discount-condition"
      >
        При покупке новой SIM–карты
      </div>
      <div
        className="mfui-product-tile-rest__old-price-wrapper"
      >
        <div
          className="mfui-product-tile-rest__old-price"
        >
          3000 ₽
        </div>
      </div>
      <div
        className="mfui-product-tile-rest__actual-price"
      >
        100 ₽ за 30 дней
      </div>
    </div>
    <div
      className="mfui-product-tile-rest__pack"
      key="10ГБ0"
    >
      10 ГБ
    </div>
    <div
      className="mfui-product-tile-rest__pack"
      key="10ГБ1"
    >
      Безлимитный интернет
    </div>
    <div
      className="mfui-product-tile-rest__messangers"
    >
      <div
        className="mfui-product-tile-rest__messangers-list"
      >
        <DropdownSocialList
          className="mfui-product-tile-rest__options-list"
          icons={
            Array [
              Object {
                "svgIcon": <test-file-stub />,
                "title": "МегаФон ТВ",
              },
            ]
          }
          maxIconNumber={4}
        />
      </div>
      <div
        className="mfui-product-tile-rest__messangers-description"
      >
        Параметры
        <br />
         
        Тариф для планшетов
      </div>
    </div>
    <div
      className="mfui-product-tile-rest__options"
    >
      <div
        className="mfui-product-tile-rest__option"
        key="Кэшбэк 20%0"
      >
        <div
          className="mfui-product-tile-rest__option-title"
        >
          Кэшбэк 20%
        </div>
        <div
          className="mfui-product-tile-rest__option-description"
        >
           
        </div>
      </div>
      <div
        className="mfui-product-tile-rest__option"
        key="Безлимитные SMS1"
      >
        <div
          className="mfui-product-tile-rest__option-title"
        >
          Безлимитные SMS
        </div>
        <div
          className="mfui-product-tile-rest__option-description"
        >
          100 ₽ за 30 дней
        </div>
      </div>
    </div>
    <div
      className="mfui-product-tile-rest__description"
    >
      description
    </div>
  </div>
  <div
    className="mfui-product-tile-rest__buy"
  >
    <Button
      className="mfui-product-tile-rest__buy-button"
      disabledColor="gray"
      downColor="dark"
      hoverColor="green"
      href="/zakaz/?tariff=choose"
      onClick={[Function]}
      padding={true}
      passiveColor="green"
      showSpinner={false}
      sizeAll="medium"
      type="button"
      width="auto"
    >
      Купить
    </Button>
    <TextLink
      className="mfui-product-tile-rest__detail-link mfui-product-tile-rest__detail-link_type_connect"
      color="blue"
      href="/tariffs/all/vybiray.html"
      onClick={[Function]}
      target="_blank"
      underlineStyle="solid"
      underlineVisibility="hover"
    >
      Подключить
    </TextLink>
  </div>
</div>
`;

exports[`<ProductTileRest /> render renders ProductTileRest with empty props 1`] = `
<div
  className="mfui-product-tile-rest"
>
  <div
    className="mfui-product-tile-rest__info"
  >
    <Header
      as="h3"
      className="mfui-product-tile-rest__header"
      color="black"
    >
      title
    </Header>
    <TextLink
      className="mfui-product-tile-rest__detail-link mfui-product-tile-rest__detail-link_type_more"
      color="blue"
      onClick={[Function]}
      target="_blank"
      underlineStyle="solid"
      underlineVisibility="hover"
    >
      Подробнее
    </TextLink>
    <div
      className="mfui-product-tile-rest__price"
    >
      <div
        className="mfui-product-tile-rest__discount-condition"
      >
        title
      </div>
      <div
        className="mfui-product-tile-rest__old-price-wrapper"
      >
        <div
          className="mfui-product-tile-rest__old-price"
        >
          value unitValue
        </div>
      </div>
      <div
        className="mfui-product-tile-rest__actual-price"
      >
        value unitValue unitExtra
      </div>
    </div>
    <div
      className="mfui-product-tile-rest__messangers"
    />
  </div>
  <div
    className="mfui-product-tile-rest__buy"
  >
    <Button
      className="mfui-product-tile-rest__buy-button"
      disabledColor="gray"
      downColor="dark"
      hoverColor="green"
      onClick={[Function]}
      padding={true}
      passiveColor="green"
      showSpinner={false}
      sizeAll="medium"
      type="button"
      width="auto"
    >
      Купить
    </Button>
    <TextLink
      className="mfui-product-tile-rest__detail-link mfui-product-tile-rest__detail-link_type_connect"
      color="blue"
      onClick={[Function]}
      target="_blank"
      underlineStyle="solid"
      underlineVisibility="hover"
    >
      Перейти на тариф
    </TextLink>
  </div>
</div>
`;
